v-next/core/test/internal/tasks/task-manager.ts

```diff
@@ -0,0 +1,908 @@
+import type { HardhatUserConfig } from "../../../src/types/config.js";
+import type { GlobalParameterMap } from "../../../src/types/global-parameters.js";
+import type { HardhatPlugin } from "../../../src/types/plugins.js";
+
+import assert from "node:assert/strict";
+import { describe, it } from "node:test";
+
+import { HardhatError } from "@nomicfoundation/hardhat-errors";
+
+import { ParameterType } from "../../../src/config.js";
+import { createHardhatRuntimeEnvironment } from "../../../src/index.js";
+import { buildGlobalParameterDefinition } from "../../../src/internal/global-parameters.js";
+import { resolvePluginList } from "../../../src/internal/plugins/resolve-plugin-list.js";
+import {
+  NewTaskDefinitionBuilderImplementation,
+  TaskOverrideDefinitionBuilderImplementation,
+} from "../../../src/internal/tasks/builders.js";
+import { TaskDefinitionType } from "../../../src/types/tasks.js";
+
+/**
+ * There is a circular dependency between the TaskManagerImplementation and the
+ * HardhatRuntimeEnvironmentImplementation. The TaskManagerImplementation needs
+ * the HardhatRuntimeEnvironmentImplementation to be created, and the
+ * HardhatRuntimeEnvironmentImplementation creates the TaskManagerImplementation.
+ *
+ * The way to test the TaskManagerImplementation is through the
+ * HardhatRuntimeEnvironmentImplementation, as it's the one that creates the
+ * TaskManagerImplementation.
+ */
+async function createHre({
+  config = {},
+  plugins = [],
+}: {
+  config?: HardhatUserConfig;
+  plugins?: HardhatPlugin[];
+  globalParameterMap?: GlobalParameterMap;
+} = {}) {
+  const resolvedPlugins = await resolvePluginList(plugins, process.cwd());
+
+  return createHardhatRuntimeEnvironment(
+    config,
+    {},
+    {
+      resolvedPlugins,
+    },
+  );
+}
+
+describe("TaskManagerImplementation", () => {
+  it("should initialize the task manager with an empty set of tasks if no plugins or tasks are provided", async () => {
+    await assert.doesNotReject(createHre());
```

Comment:
This test isn't really testing what its description does. At the same time, I don't think we have a way to actually test it, which makes me think that the `TaskManager` interface is incomplete. Maybe it should be able to either list all the tasks, or at least all the root tasks.

A similar problem will appear when printing the help messages.

Reply:
Maybe add a `getTopLevelTasks()` in a separate PR?